pkg/crypto/x509/x509.go

```diff
@@ -538,6 +539,80 @@ func (p *PEMEncodedCertificateAndKey) MarshalYAML() (interface{}, error) {
 	return aux, nil
 }
 
+// NewCertficateAndKey generates a new key and certificate signed by a CA.
+//
+//nolint: gocyclo
+func NewCertficateAndKey(crt *x509.Certificate, key interface{}, setters ...Option) (p *PEMEncodedCertificateAndKey, err error) {
+	opts := NewDefaultOptions(setters...)
+
+	var (
+		c        *Certificate
+		k        interface{}
+		pemBytes []byte
+	)
+
+	if opts.RSA {
+		k, err = NewRSAKey()
+		if err != nil {
+			return nil, err
+		}
+
+		pemBytes = k.(*RSAKey).KeyPEM
+	} else {
+		k, err = NewKey()
+		if err != nil {
+			return nil, err
+		}
+
+		pemBytes = k.(*Key).KeyPEM
+	}
+
+	block, _ := pem.Decode(pemBytes)
+	if block == nil {
+		return nil, errors.New("failed to decode admin key pem")
+	}
+
+	var priv interface{}
+	if opts.RSA {
+		priv, err = x509.ParsePKCS1PrivateKey(block.Bytes)
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		priv, err = x509.ParseECPrivateKey(block.Bytes)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	csr, err := NewCertificateSigningRequest(priv, setters...)
+	if err != nil {
+		return nil, err
```

Comment:
errors in this function might benefit from any kind of wrapping (fmt.Errorf with %w or errors.Wrap), as sometimes it's hard to understand which part of the long flow returned an error

Reply:
Agree 👍